benchmarks/src/test/java/org/apache/druid/timeline/DimensionRangeShardSpecBenchmark.java

```diff
@@ -0,0 +1,146 @@
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Fork(value = 1, jvmArgsAppend = {"-Xms128m", "-Xmx128m", "-XX:+UseG1GC"})
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@BenchmarkMode({Mode.AverageTime})
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@OperationsPerInvocation(3) // 3 shard specs per invocation
+@SuppressWarnings("UnstableApiUsage") // Range, RangeSet are unstable APIs
+public class DimensionRangeShardSpecBenchmark
+{
+  private final Map<String, RangeSet<String>> domainSinglePrunable =
+      ImmutableMap.<String, RangeSet<String>>builder()
+                  .put("country", ImmutableRangeSet.of(Range.singleton("India")))
+                  .build();
+
+  // Initialized in @Setup method.
+  private Map<String, RangeSet<String>> domain5kPrunable;
+
+  private final Map<String, RangeSet<String>> domainSingleNonPrunable =
+      ImmutableMap.<String, RangeSet<String>>builder()
+                  .put("country", ImmutableRangeSet.of(Range.singleton("Shanghai")))
+                  .build();
+
+  // Initial segment (null -> values)
+  private final DimensionRangeShardSpec shardSpec0 = new DimensionRangeShardSpec(
+      Arrays.asList("country", "city"),
+      new StringTuple(new String[]{null, null}),
+      new StringTuple(new String[]{"Germany", "Munich"}),
+      0,
+      4
+  );
+
+  // Middle segment (values -> other values)
+  private final DimensionRangeShardSpec shardSpec1 = new DimensionRangeShardSpec(
+      Arrays.asList("country", "city"),
+      new StringTuple(new String[]{"Germany", "Munich"}),
+      new StringTuple(new String[]{"United States", "New York"}),
+      1,
+      4
+  );
+
+  // End segment (values -> null)
+  private final DimensionRangeShardSpec shardSpec2 = new DimensionRangeShardSpec(
+      Arrays.asList("country", "city"),
+      new StringTuple(new String[]{"United States", "New York"}),
+      new StringTuple(new String[]{null, null}),
+      2,
+      4
+  );
+
+  @Setup
+  public void setUp()
+  {
+    NullHandling.initializeForTests();
+
+    final Set<String> strings5k = new HashSet<>();
+
+    final Random random = new Random(0); // Random... ish.
+    for (int i = 0; i < 5000; i++) {
+      // Uppercase so the strings end up in different shardSpecs
+      final String s = StringUtils.format(
+          "%s%s",
+          String.valueOf((char) ('A' + random.nextInt(26))),
+          String.valueOf(random.nextInt())
+      );
+      strings5k.add(s);
+    }
+
+    final RangeSet<String> rangeSet5k = new InDimFilter("country", strings5k).getDimensionRangeSet("country");
```

Comment:
So this range set is effectively a union of many singletons, which is what we get with an IN filter. But I suppose that is what we are solving for here.

Perhaps we could add other benchmarks here later that also test on range sets that contain continuous ranges.

Reply:
Yeah, for sure. This patch came out of trying to optimize a scenario that involved a large IN filter.